cmd/kubeadm/app/phases/controlplane/users_linux.go

```diff
@@ -0,0 +1,260 @@
+	if err := os.Chown(filepath, int(uid), int(gid)); err != nil {
+		return err
+	}
+	if err := os.Chmod(filepath, fs.FileMode(permissions)); err != nil {
+		return err
+	}
+	return nil
+}
+
+func UpdateFileOwnership(filepath string, uid int64, gid int64, permissions uint32) error {
+	return UpdateFileOwnershipFn(filepath, uid, gid, permissions)
+}
+
+var UpdateDirecotryOwnershipFn = func(directorypath string, uid int64, gid int64, permissions string) error {
+	exec := exec.New()
+	cmd := exec.Command("chown", "-R", fmt.Sprintf("%d:%d", uid, gid), directorypath)
+	buff := bytes.Buffer{}
+	cmd.SetStderr(&buff)
+	if err := cmd.Run(); err != nil {
+		return errors.New("failed to change directory owner: " + err.Error() + " : " + buff.String())
+	}
+	cmd = exec.Command("chmod", "-R", permissions, directorypath)
+	buff = bytes.Buffer{}
+	cmd.SetStderr(&buff)
+	if err := cmd.Run(); err != nil {
+		return errors.New("failed to change directory permission: " + err.Error() + " : " + buff.String())
+	}
+	return nil
+}
+
+func UpdateDirecotryOwnership(directorypath string, uid int64, gid int64, permissions string) error {
+	return UpdateDirecotryOwnershipFn(directorypath, uid, gid, permissions)
+}
+
+func getControlPlaneUserId(username string) (int64, bool, error) {
+	file, err := os.Open("/etc/passwd")
+	defer file.Close()
+	if err != nil {
+		return 0, false, err
+	}
+	scanner := bufio.NewScanner(file)
+	scanner.Split(bufio.ScanLines)
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.HasPrefix(line, username) {
+			uid, err := strconv.ParseInt(strings.Split(line, ":")[2], 10, 64)
+			if err != nil {
+				return 0, false, err
+			}
+			return uid, true, nil
+		}
+	}
+	return 0, false, nil
+}
+
+func getControlPlaneGroupId(groupname string) (int64, bool, error) {
+	file, err := os.Open("/etc/group")
+	defer file.Close()
+	if err != nil {
+		return 0, false, err
+	}
+	scanner := bufio.NewScanner(file)
+	scanner.Split(bufio.ScanLines)
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.HasPrefix(line, groupname) {
+			gid, err := strconv.ParseInt(strings.Split(line, ":")[2], 10, 64)
+			if err != nil {
+				return 0, false, err
+			}
+			return gid, true, nil
+		}
+	}
+	return 0, false, nil
+}
+
+// createControlPlanUser creates a user. It does not check if the user
+// exists before creating it.
+func createControlPlanGroup(groupname string) error {
+	exec := exec.New()
+	cmd := exec.Command("groupadd", "-r", groupname)
+	buff := bytes.Buffer{}
+	cmd.SetStderr(&buff)
+	if err := cmd.Run(); err != nil {
+		return errors.New(fmt.Sprintf("failed to add group %s: %s : %s", groupname, err.Error(), buff.String()))
+	}
+	return nil
+}
+
+// createControlPlanUser creates a user. It does not check if the user
+// exists before creating it.
+func createControlPlanUser(username string) error {
+	exec := exec.New()
+	cmd := exec.Command("useradd", "-r", "-s", "/bin/false", username)
```

Comment:
shell out is problematic because it is making some assumptions around Linux distros and what they package.

what if useradd/del is not present in some trimmed container distros?
can we avoid using /bin/false too?